client/jetpack-cloud/controller.ts

```diff
@@ -0,0 +1,178 @@
+		id: getSiteId( getState(), siteId ),
+		slug: getSiteSlug( getState(), siteId ),
+	} ) );
+};
+
+/**
+ * Store site id in state, and add site to the list of most recent sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ */
+const selectSite = ( context: PageJS.Context, siteId: number ): void => {
+	const { dispatch } = context.store;
+
+	dispatch( setSelectedSiteId( siteId ) );
+	updateRecentSitesPreferences( context );
+};
+
+/**
+ * Handle case when path contains no site slug. If the current user has only
+ * one site, we try to use this site for context. Otherwise, the user must
+ * select one of their sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+const siteSelectionWithoutFragment = ( context: PageJS.Context, next: () => void ): void => {
+	const { getState, dispatch } = context.store;
+	const currentUser = getCurrentUser( getState() ) as UserData;
+	const hasOneSite = currentUser?.visible_site_count === 1;
+
+	// User has only one site
+	if ( hasOneSite ) {
+		const primarySiteId = getPrimarySiteId( getState() ) as NonNullable< number >;
+		const primarySiteSlug = getSiteSlug( getState(), primarySiteId );
+
+		// Site is already available in state
+		if ( primarySiteSlug ) {
+			// Redirect to the single site context
+			redirectToPrimary( context, primarySiteSlug );
+		} else {
+			fetchSite( context, primarySiteId ).then( ( { slug } ) => {
+				if ( slug ) {
+					// Redirect to the single site context
+					redirectToPrimary( context, slug );
+				} else {
+					// TODO: display error
+				}
+			} );
+		}
+	} else {
+		// Show all sites
+		dispatch( setAllSitesSelected() );
+		next();
+	}
+};
+
+/**
+ * Select site when path contains a site slug.
+ *
+ * @param {string} siteFragment Parsed site slug
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+const siteSelectionWithFragment = (
+	siteFragment: string,
+	context: PageJS.Context,
+	next: () => void
+): void => {
+	const { getState } = context.store;
+	const siteId = getSiteId( getState(), siteFragment );
+
+	// Site is already available in state
+	if ( siteId ) {
+		selectSite( context, siteId );
+		next();
+	} else {
+		fetchSite( context, siteFragment ).then( ( { id } ) => {
+			if ( id ) {
+				selectSite( context, id );
+				next();
+			} else {
+				const allSitesPath = addQueryArgs(
+					{ site: siteFragment },
+					sectionify( context.path, siteFragment )
+				);
+
+				page.redirect( allSitesPath );
+			}
+		} );
+	}
+};
+
+/**
+ * Parse site slug from path and call the proper middleware.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+export function cloudSiteSelection( context: PageJS.Context, next: () => void ): void {
+	const siteFragment = parseSiteFragment( context );
+
+	// TODO: handle no Jetpack site
```

Comment:
As a note, this is going to be handled here:
1164141197617539-as-1176558595787257

For now it'd be okay to redirect to the current no sites page.